entsoe-balances-adjustment/src/main/java/com/powsybl/entsoe/balances_adjustment/util/CgmesVoltageLevelsArea.java

```diff
@@ -0,0 +1,130 @@
+        return network.getDanglingLineStream()
+                .filter(this::isAreaBorder)
+                .filter(dl -> dl.getTerminal().getBusView().getBus() != null && dl.getTerminal().getBusView().getBus().isInMainSynchronousComponent())
+                .filter(dl -> dl.getExtension(CgmesDanglingLineBoundaryNode.class) == null || !dl.getExtension(CgmesDanglingLineBoundaryNode.class).isHvdc())
+                .filter(dl -> {
+                    if (area != null && (!area.getTerminals().isEmpty() || !area.getBoundaries().isEmpty())) {
+                        return area.getTerminals().stream().anyMatch(t -> t.getConnectable().getId().equals(dl.getId())) || area.getBoundaries().stream().anyMatch(bd -> bd.getConnectable().getId().equals(dl.getId()));
+                    }
+                    return true;
+                })
+                .filter(dl -> {
+                    if (excludedXnodes != null) {
+                        return excludedXnodes.stream().noneMatch(xnodeCode -> dl.getUcteXnodeCode().equals(xnodeCode));
+                    }
+                    return true;
+                })
+                .collect(Collectors.toList());
+    }
+
+    private List<Branch<?>> createBranchesCache(Network network, CgmesControlArea area, List<String> excludedXnodes) {
+        return network.getLineStream()
+                .filter(this::isAreaBorder)
+                .filter(b -> b.getTerminal1().getBusView().getBus() != null && b.getTerminal1().getBusView().getBus().isInMainSynchronousComponent()
+                        && b.getTerminal2().getBusView().getBus() != null && b.getTerminal2().getBusView().getBus().isInMainSynchronousComponent())
+                .filter(b -> b.getExtension(CgmesLineBoundaryNode.class) == null || !b.getExtension(CgmesLineBoundaryNode.class).isHvdc())
+                .filter(b -> !b.hasProperty("isHvdc")) // necessary as extensions of merged lines are not well handled
+                .filter(b -> {
+                    if (area != null && (!area.getTerminals().isEmpty() || !area.getBoundaries().isEmpty())) {
+                        if (b instanceof TieLine) {
+                            return area.getTerminals().stream().anyMatch(t -> b.getId().contains(t.getConnectable().getId()))
+                                    || area.getBoundaries().stream().anyMatch(bd -> b.getId().contains(bd.getConnectable().getId()));
+                        } else {
+                            return area.getTerminals().stream().anyMatch(t -> b.getId().contains(t.getConnectable().getId()));
+                        }
+                    }
+                    return true;
+                })
+                .filter(b -> {
+                    if (b instanceof TieLine && excludedXnodes != null) {
+                        TieLine tl = (TieLine) b;
+                        return excludedXnodes.stream().noneMatch(xnodeCode -> tl.getUcteXnodeCode().equals(xnodeCode));
```

Comment:
The excluded Xnodes are only on TieLine?

Reply:
Yes and except on weird cases, we should only have tie lines here anyway (I described other cases in order not to have the class crash if it happens but it shouldn't happen in my opinion)